keps/sig-api-machinery/2839-in-use-protection/README.md

```diff
@@ -0,0 +1,682 @@
+Finalizers block the _completion_ of a delete operation, but they do not prevent the deletion from starting.
+Once started, a delete _will_ complete and can not be aborted.
+
+However, Finalizer may not be appropriate for some resource protection use cases, because it doesn't prevent other controllers from executing their pre-deletion actions.
+The order of execution of pre-deletion actions across finalizers is not defined.
+As a result, other controllers will execute their pre-deletion actions while a finalizer for protection still exists.
+
+This KEP aims to provide a generic mechanism for protection that solves the above issues with Finalizers, because such use cases are also found for protecting secrets and configmaps in [KEP-2639](https://github.com/kubernetes/enhancements/pull/2640).
+
+Potential use cases are:
+- Protect `PersistentVolume` while being used by `PersistentVolumeClaim` (Replace the background logic for pv-protection)
+- Protect `PersistentVolumeClaim` while being used by `Pod` (Replace the background logic for pvc-protection)
+- Protect `Secret` while being used by `Pod`, `PersistentVolume`, `VolumeSnapshotContent` (Secret protection above)
+- Protect `ConfigMap` while being used by `Pod`
+- Protect `StorageClass` while being used by `PersistentVolume`
+- Protect `VolumeSnapshot` while being used by `VolumeSnapshotContent`
+- Protect `VolumeSnapshotClass` while being used by `VolumeSnapshotContent`
+- Protect `User`, `Group`, `ServiceAccount` and `Role` from `RoleBinding`
+- Protect `User`, `Group`, `ServiceAccount` and `ClusterRole` from `ClusterRoleBinding`
+- Protect resources that users marked as in-use
+
+### Goals
+
+- Provide a generic mechanism that prevents a resource from being deleted when it shouldn't be deleted
+
+### Non-Goals
+
+- Provide a specific mechanism to decide which particular objects should be prevented from being deleted when other particular objects exist
+- Provide a generic mechanism that prevents a resource from being updated when it shouldn't be updated
+- Implement any of the above potential use cases
+
+## Proposal
+
+A new field `Liens` to mark the object not to be deleted is introduced in object Metadata.
+Deletion requests for resources with non-empty `Liens` will be blocked by a newly introduced validation in api-server.
+
+### User Stories (Optional)
+
+#### Story 1
+
+Another controller to protect Secret from deletion watches all Secrets and their potential user objects, like `Pod`, `PersistentVolumes`, and `VolumeSnapshotContent`. Once it find the Secret is used by one of the objects, it updates the Secret's `Liens` and ask this mechanism to block deletion request for the secret.
+
+#### Story 2
+
+A user sets `Liens` to an object to be protected by this mechanism. For example, when a user knows that `Namespace` X shouldn't be erroneously deleted while `Deployment` Y exists for his/her particular use case, the user can set such a protection.
+
```

Comment:
add an accidental deletion protection user story? E.g. User knows this is an important object, and that e.g. terraform likes to delete and recreate objects, so the user sets a lien so that an accidental deletion will fail.

Reply:
Replaced the story 2 with it (because references to the using objects are no longer managed by lien, so the old one is covered by the new one).